syntax = "proto3";

package chat;

// Task hint
// The google packages should be available to you when compiling this proto
// If they are are not and it becomes an issues, remove the import and replace the type with a unix timestamp (uint64)
import "google/protobuf/timestamp.proto";

// Task hint
// options for other languages might be needed
option go_package="github.com/brunsgaard/lets-chat";

// Service definition for a chat servier
service Chat {
    // Login is called by a user that attempts to join the server
    rpc Login(LoginRequest) returns (LoginResponse) {}
    // Logout is called when a user want to leave the chat server
    rpc Logout(LogoutRequest) returns (LogoutResponse) {}
    // Stream is a bidirectional stream, the user sends chat messages to the
    // server and recieves a stream og events happening on the server.
    // IMPORTANT: Use grpc metadata to include the token from LoginResponse
    // when calling Stream to identify the caller, the metadata header is
    // called `x-chat-token` in the provided example implementation.
    rpc Stream(stream StreamRequest) returns (stream StreamResponse) {}
}

message LoginRequest {
    // password is the password for the server (not the user). Everybody with
    // the server password can join.
    string password = 1;
    // The name of the user that wants to login
    string name     = 2;
}

message LoginResponse {
    // A token (session) is uniq and randomly generated by the server, the
    // server uses the token to keep track of users logged in. The token should
    // be send as metadata in the Stream call.
    string token = 1;
}

message LogoutRequest {
    // Send the token so the server know what session/stream to terminate.
    string token = 1;
}

message LogoutResponse {}

message StreamRequest {
    string message = 2;
}

// StreamResponse is a stream of events coming back from the server, the event type can be
// A user logded in
// A user logged out
// A user posted a message
// The server is shutting down
message StreamResponse {

    // replace with uint64 is import you see import issues with google packages
    google.protobuf.Timestamp timestamp = 1;

    oneof event {
        Login    client_login    = 2;
        Logout   client_logout   = 3;
        Message  client_message  = 4;
        Shutdown server_shutdown = 5;
    }

    message Login {
        string name = 1;
    }

    message Logout {
        string name = 1;
    }

    message Message {
        string name    = 1;
        string message = 2;
    }

    message Shutdown {}
}